File: src/controller/webhook.go
package controller

import (
	"encoding/json"
	"io"
	"log"
	"net/http"

	"github.com/Rowentey/teybot/src/model"
	"github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"
)

func WebhookHandler(w http.ResponseWriter, r *http.Request, b *bot.Bot) {
	// Only allow POST requests
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		log.Printf("Failed to read request body: %v", err)
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	const maxLog = 1024
	if len(body) > maxLog {
		log.Printf("Debug body (truncated %d bytes): %s...", len(body), string(body[:maxLog]))
	} else {
		log.Printf("Debug body: %s", string(body))
	}

	var msgReq model.MessageRequest
	if err := json.Unmarshal(body, &msgReq); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Validate request
	if msgReq.ChatID == 0 || msgReq.Text == "" {
		http.Error(w, "Missing required fields: chat_id or text", http.StatusBadRequest)
		return
	}

	params := &bot.SendMessageParams{
		ChatID: msgReq.ChatID,
		Text:   msgReq.Text,
	}

	if msgReq.MessageThreadID != 0 {
		params.ReplyParameters = &models.ReplyParameters{
			MessageID: msgReq.MessageThreadID,
		}
	}

	// Send message using the bot
	if _, err := b.SendMessage(r.Context(), params); err != nil {
		log.Printf("Error sending message: %v", err)
		http.Error(w, "Failed to send message", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
}

File: src/routes/webhook.ts
import { getBotAndHandler } from "../telegram";
import type { Env } from "../types";

interface MessageRequest {
	chat_id: number;
	message_thread_id?: number;
	title: string;
	message: string;
}

function jsonResponse(status: number, body: unknown): Response {
	return new Response(JSON.stringify(body), {
		status,
		headers: { "content-type": "application/json" },
	});
}

function escapeMarkdownV2(input: string): string {
	const special = /[_*\[\]()~`>#+\-=|{}.!]/g;
	return input.replace(special, "\\$&");
}

export async function handleWebhookSend(
	request: Request,
	env: Env,
): Promise<Response> {
	let payload: MessageRequest;
	try {
		payload = (await request.json()) as MessageRequest;
	} catch {
		return jsonResponse(400, { error: "Invalid request body" });
	}

	if (!payload.chat_id) {
		return jsonResponse(400, { error: "Missing required field: chat_id" });
	}
	if (!payload.title) {
		return jsonResponse(400, { error: "Missing required field: title" });
	}
	if (!payload.message) {
		return jsonResponse(400, { error: "Missing required field: message" });
	}

	const { bot } = getBotAndHandler(env);
	const text = `*${escapeMarkdownV2(payload.title)}*\n\n${escapeMarkdownV2(payload.message)}`;

	try {
		await bot.api.sendMessage(payload.chat_id, text, {
			parse_mode: "MarkdownV2",
			message_thread_id: payload.message_thread_id,
		});
	} catch {
		return jsonResponse(500, { error: "Failed to send message" });
	}

	return new Response(null, { status: 200 });
}
